<h1 style="margin-bottom: 0.4em; text-align: center;">
    <b>Lecture 03A</b><br>
    Genome assembly
</h1>
<h2 style="margin-top: 0.0em; text-align: center;">
    Foundations
</h2>
<p style="text-align: center;">
    <b>Date:</b> Jan 21, 2024
</p>

This lecture introduces genome assembly as a critical step in the genomics pipeline, transitioning from raw sequencing reads to assembled DNA sequences.
Students will explore the differences between reference-based and de novo assembly and understand the challenges posed by repeats, sequencing errors, and low coverage.
The session concludes with an overview of the outputs generated by assembly tools like SPAdes and a discussion of quality assessment methods using tools such as QUAST and Bandage.

## Learning objectives

After today, you should have a better understanding of:

1.  Where genome assembly fits in the genomics pipeline.
2.  Reference-based and de novo genome assembly.
3.  Challenges like repeats and sequencing errors.
4.  Outputs of genome assembly tools.
5.  Tools for assessing assembly quality.

## Supplementary material

Relevant content for today's lecture.

-   [Genome assembly](https://omics.crumblearn.org/genomics/assembly/)
-   [Assembly concepts](https://omics.crumblearn.org/genomics/assembly/concepts/) and nested content

<!-- ## Presentation

-   **View:** [slides.com/aalexmmaldonado/biosc1540-l04](https://slides.com/aalexmmaldonado/biosc1540-l04)
-   **Live link:** [slides.com/d/KOit8yE/live](https://slides.com/d/KOit8yE/live)
-   **Download:** [biosc1540-l04.pdf](/lectures/04/biosc1540-l04.pdf)

<iframe src="https://slides.com/aalexmmaldonado/biosc1540-l04/embed?byline=hidden&share=hidden" width="100%" height="600" title="BIOSC 1540: Lecture 04" scrolling="no" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe> -->

<!--

- Students were not really engaged with hands-on practice. I guess this is fine since they can practice it later.
- I ended early, so if I take out the activities, I certainly have time to add in a galaxy activity or the OLC.
- I had a good question about why we take k-mers, then make k-1 mers for nodes. Why not just straightly make k-1 mer nodes.
- I'm not sure it was useful to include the error correction.
- I think I need to add in a little more information about how actual assemblers build contigs from de bruijn graphs. For example, graph traversal using depth-first earch or Eulerian path algorithm. Handling branches, handling repeats. Maybe we should talk about how spades does it.
- I'm also confused on if graphs need to have Eulerian walks.
- I'm not sold on the notation of k mers for length of nodes or edges.

 -->
